Add Link's House, ToT, Windmill and Overworld entrances to ER

Roman971 opened the pull request:
This PR changes and adds options to the Entrance Shuffle setting (aka ER): 

- **'Dungeons Only'**: Not changed at all.

- **'Simple Indoors'** (formerly 'All Indoors'): This is exactly the same as the original 'All Indoors' option, just renamed, and it is kept as a separate option from the new 'All Indoors' option (described below), for players who want to only shuffle entrances that don't majorly change access to overworld areas.

- **'All Indoors'**: Shuffle all Indoor entrances, including Link's House, Temple of Time and the Windmill. This basically extends the 'Simple Indoors' option to add those special places which can greatly change routing and how the world is accessed.

- **'All Indoors & Overworld'**: Shuffle all entrances from the 'All Indoors' option, along with Overworld loading zones between areas, shuffled in a new separate pool. Owl drop positions are also randomized among this new pool, as extra one-way entrances.

Along with those changes, Entrance Shuffle should now fully support the plandomizer functionality, which means it is now possible to set any entrance from a distribution file if necessary.

An entrance playthrough is also added at the end of spoiler files to help with debugging/reading seeds which include shuffled entrances. This playthrough can still be improved but I believe it can already help quite a bit to understand ER seeds, especially when overworld entrances are shuffled.

I will soon make a page on the Wiki to provide all the information missing from the setting tooltip since there is too much to say about all ER options for one tooltip on the GUI. I also want to add a list of ER quirks when it comes to the logic, so I will try to do that as soon as I have time.

_Internal implementation things to note:_
The current shuffling algorithm can definitely still be improved (especially in terms of performance), but successfully generated seeds should be fine. And generation fails should be rare but are still possible.

Some notable flaw in the current logic is the fact that the adult trade quest currently requires 'Prelude of Light' in order to be completed in logic when 'All Indoors' or 'All Indoors & Overworld' options are selected (to avoid potential logic issues due to items reverting on savewarp). This should however become irrelevant when the trade quest rework happens, which I hope will be the case soon.

Castle Town Entrance <-> Hyrule Field loading zones are currently not randomized because they trigger fog glitches, so until fog glitch is handled on the rom side (hopefully very soon), they are kept vanilla.

Entrance placement currently guarantees at least one of either Kokiri Forest or Kakariko Village are reachable from the start of the seed without items, and that at least one time passing scene is reachable as both ages. These forced placements could theoretically be removed but that would severely complexify the logic and make it much slower in general. In the end they add some small bias towards entrance placement but I think it's fine, at least for now, since it only matters for the first few reachable entrances anyway.

r0bd0g commented:
Can you check that playing Sun's at a Gossip for a Fairy doesn't just reload the area? It might be that you can only get Fairies like that in places where playing Sun's doesn't reload.

Lost Woods->Lost Woods Bridge
Hovers Boots and Longshot should both be acceptable means of getting over there. Actually adult can do it with no items, but that's trick territory.

I didn't see the Ganon Scrub Room Fairies? Maybe I just didn't see them.

Could behind Kak gate -> Kak be true if the gate is there and visibly blocking you, even though it doesn't have collision?

Do you need to check for child to get the fish in domain? I don't know for sure how all this works uh, at some point I'm going to want to figure it out.

r0bd0g commented:
You changed it to only be used where time passes however, not all places where time doesn't pass reload when you play sun's song.

Roman971 commented:
Thanks for taking a look at the logic. I pushed changes that should fix everything you mentioned apart from the Kakariko gate thing, I'm still unsure myself if that should be in logic or not. 

As you mentioned, it doesn't seem to be possible to get fairies from gossip stones when using Sun's Song in a scene where time doesn't pass, so I added a check for that.
The check for being child to get a fish in zora's domain and ganon's castle scrub fairies were indeed missing so I added them. 

Along with longshot and hover boots as ways to get on the bridge, I also added longshot as the only way to go down from the bridge and reach the rest of LW. I think that should be fine as well.

r0bd0g commented:
You changed it to only be used where time passes however, not all places where time doesn't pass reload when you play sun's song. You should test in DC. And I Grottos I forget if Sun's reloads it?

Yeah, I forgot about longshotting off the bridge, I guess. :p

Roman971 commented:
Right, I had forgotten about the gossip stone in DC. Playing sun's song also doesn't seem to trigger a reload in grottos, I had no idea about that though.

I guess it can't be handled by looking at the time_passes property then. That's unfortunate but I'll try to find a decent alternative.

flagrama commented:
It looks like the back of Impa's house is set to "True" regardless of age already? If so that means there is already a one-way collision in logic doesn't it? In which case either closed Kak gate should be in logic to pass through from the back or Impa's house needs to be set to True as child but require hookshot (not sure if anything else?) as Adult.

```
"Impas House Back": "True",
```
```
{
        "region_name": "Impas House Back",
        "locations": {
            "Impa House Freestanding PoH": "True",
            "Impas House Cow": "can_play(Eponas_Song)"
        }
},
```

(Also in the case of changing Impa's House Back, do you need to be in the back to get milk/item from the cow, or does it also work from the front of Impa's House too?)

r0bd0g commented:
Impa's House... That's a one-way collision that I really think should not be in logic by default, maybe at all.

Roman971 commented on ItemPool.py:
```diff
@@ -601,6 +613,13 @@ def generate_itempool(world):
         world.get_location(location).locked = True
 
+    drop_locations = list(filter(lambda loc: loc.type == 'Drop', world.get_locations()))
+    for location, item in droplocations.items():
```
You're right, I'm not sure why I didn't think of that while writing this. I'll change it.

Roman971 commented on Main.py:
```diff
@@ -541,17 +543,25 @@ def create_playthrough(spoiler):
         # Not collecting while the generator runs means we only get one sphere at a time
         # Otherwise, an item we collect could influence later item collection in the same sphere
         collected = list(playthrough.iter_reachable_locations(required_locations))
+        accessed_entrances = [entrance for world in worlds for entrance in world.get_shuffled_entrances() 
```
I wanted to do something similar at first but I don't know a clean way to do that in Python. I don't think there is any built-in way to partition/split a list based on a predicate, is there?
I would like to do something like this but i can't find a decent way to do it in Python:
```python
accessed, remaining = partition(lambda entrance: can_reach(entrance), remaining)
```
I guess it can be done with 2 consecutive filters but I'm curious to know if there is a better way.

Zannick replied:
Not really a one-line partitioner, but I would do
```py
accessed = set(filter(can_reach, remaining))
remaining -= accessed
```

Roman971 commented on Main.py:
```diff
@@ -541,17 +543,25 @@ def create_playthrough(spoiler):
     spoiler.playthrough = OrderedDict((str(i + 1), {location: location.item for location in sphere}) for i, sphere in enumerate(collection_spheres))
 
+    if worlds[0].entrance_shuffle != 'off':
+        spoiler.entrance_playthrough = OrderedDict((str(i + 1), [entrance for entrance in sphere]) for i, sphere in enumerate(entrance_spheres))
```
I also think it could be better to include it in the normal playthrough, but since it doesn't only include useful entrances, I left it in a separate playthrough for now. 
I wanna try to have it only include entrances actually useful to items collected in the playthrough, but I'm not sure how to do that without losing too much performance.
Once it works this way, it will probably make more sense to include it in the item playthrough, sorted last as you mentioned.

Zannick commented:
Having played through a seed with entrance_shuffle=all, I've gotta say, it was a blast and you've done a great job. I have to agree with your point above on the entrance playthrough; it is incredibly dense with all the entrances found. I had only three issues, though I'm not certain if the first two relate to the entrance randomization.

1.   I had some trouble with the master sword becoming unequipped. The equipment tab on the pause screen still showed it equipped, but the HUD button did not show it, and B didn't do anything. I don't know what was triggering that. (I never found Kokiri Sword, and I couldn't reproduce it by using Epona.)
2.   I triggered the [fog glitch](http://www.jaytheham.com/zcw/Ocarina_of_Time_Miscellaneous_Glitches_-_Fogged) as adult somehow, after teleporting into Graveyard, Graveyard -> Kakariko into Lost Woods from Zora River (infinite visibility), LW -> GC Woods Warp into Gerudo Valley Far Side (zero visibility), Gerudo Valley -> Hyrule field into Darunia (fog cleared), GC -> DMT into Lon Lon Ranch (infinite visibility), LLR -> HF back into Goron City, GC Woods Warp -> LW into the Market and no fog after that.
3.   I discovered that if you ride Epona into a loading zone that takes you to a place that can't have Epona, when you next arrive in a place where you can, you'll be riding Epona. This is an issue for three exits that I've found: Hyrule Field from Kakariko, where she gets stuck on the stairs (but Link can dismount); teleporting into Lake Hylia (she can't step down from the platform, but Link can dismount); and Lake Hylia from Zora's Domain, where Link and Epona are set above the world and can't move (Link can shoot arrows but can't dismount).

teallen1996 commented:
Also playing a seed on this and want to quickly enquire something. Is save warping a thing that is to be expected to need to be done in order to move about. If not, then I may have come across a seed where a 'softlock' is produced (can free self by save warping, but that is the question)

As a Child, going from death mountain to Goron City takes me to the Lost Woods Dive point (Silver Scale Required to get back), and in terms of the lost woods exits:

To Kokiri Forest takes me to Gerudo Fortress (Valley entrance side) (Cannot get to another exit)
To Sacred Forest Meadow takes me to Hyrule Castle Grounds (A place where there are no other exits)
To Goron City takes me to Gerudo Fortress (Wasteland side behind the gate) (Cannot get to another exit)

All of this means that the only way to get back out to a non dead end location is the zora river dive point which needs the silver scale which I didn't have. As mentioned, this is exitable using savewarp, but not sure if this was an intended behaviour to be required.

This is the seed if anyone is interested.
":version":              "4.9.13 f.LUM",
  "file_hash":             ["Bottled Milk", "Hammer", "Bottled Fish", "Longshot", "Hover Boots"],
  ":seed":                 "U8NAJ8KGTP",
  ":settings_string":      "AJRAN6HAS9NXMDJEFASSBENAKVUAS3LDVDA",

Zannick commented:
Yes, save warping is implicitly expected to be possible in all seeds, ER is just much more reliant on it.

Color-Printer commented:
Shouldn't the waterfall/river warp from gerudo valley to lake hylia act like an extra one-way warp like the owls? Or is it not shuffled for technical reasons?

Roman971 commented:
The main issue with changing the valley -> lake warp to work like owls is that it would end up permanently removing one of the main ways to get to Lake Hylia, whereas owls are already bonus ways to get around the world.
